File: frontend/src/types/insight.ts
export interface SpendingSummary {
  total_spent: number;
  total_income: number;
  net_balance: number;
  top_categories: CategorySpending[];
  comparison: {
    previous_period: number;
    change_percentage: number;
    trend: 'up' | 'down' | 'stable';
  };
}

export interface CategorySpending {
  category_id: string;
  category_name: string;
  category_color: string;
  amount: number;
  percentage: number;
  transaction_count: number;
}

export interface AIInsight {
  id: string;
  type: 'tip' | 'warning' | 'achievement' | 'prediction';
  title: string;
  description: string;
  priority: 'low' | 'medium' | 'high';
  created_at: string;
}

export interface ChatMessage {
  id: string;
  role: 'user' | 'assistant';
  content: string;
  timestamp: string;
}

export interface SpendingPrediction {
  period: string;
  predicted_amount: number;
  confidence: number;
  breakdown: {
    category_id: string;
    category_name: string;
    predicted_amount: number;
  }[];
}

File: backend/app/models/insight.py
from pydantic import BaseModel
from typing import Literal
from datetime import datetime


class CategorySpending(BaseModel):
    """Category spending breakdown."""

    category_id: str
    category_name: str
    category_color: str
    amount: float
    percentage: float
    transaction_count: int


class SpendingComparison(BaseModel):
    """Spending comparison with previous period."""

    previous_period: float
    change_percentage: float
    trend: Literal["up", "down", "stable"]


class SpendingSummary(BaseModel):
    """Spending summary model."""

    total_spent: float
    total_income: float
    net_balance: float
    top_categories: list[CategorySpending]
    comparison: SpendingComparison


class AIInsight(BaseModel):
    """AI generated insight."""

    id: str
    type: Literal["tip", "warning", "achievement", "prediction"]
    title: str
    description: str
    priority: Literal["low", "medium", "high"]
    created_at: datetime


class ChatMessage(BaseModel):
    """Chat message model."""

    role: Literal["user", "assistant"]
    content: str


class ChatRequest(BaseModel):
    """Chat request model."""

    message: str
    history: list[ChatMessage] = []


class ChatResponse(BaseModel):
    """Chat response model."""

    message: str
    timestamp: datetime


class PredictionBreakdown(BaseModel):
    """Prediction breakdown by category."""

    category_id: str
    category_name: str
    predicted_amount: float


class SpendingPrediction(BaseModel):
    """Spending prediction model."""

    period: str
    predicted_amount: float
    confidence: float
    breakdown: list[PredictionBreakdown]
